src/primitives/transaction.h

```diff
@@ -351,11 +356,36 @@ class CTransaction
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        READWRITE(*const_cast<int32_t*>(&this->nVersion));
+        if (ser_action.ForRead()) {
+            // When deserializing, unpack the 4 byte header to extract fOverwintered and nVersion.
+            uint32_t header;
+            READWRITE(header);
+            *const_cast<bool*>(&fOverwintered) = header >> 31;
+            *const_cast<int32_t*>(&this->nVersion) = header & 0x7FFFFFFF;
+        } else {
+            if (fOverwintered) {
+                // When serializing v3, assemble the 4 byte header from fOverwintered and nVersion
+                uint32_t header = (fOverwintered << 31) | this->nVersion;
+                READWRITE(*const_cast<uint32_t*>(&header));
+            } else {
+                // When serializing v1 and v2, the 4 byte header is nVersion
+                READWRITE(*const_cast<int32_t*>(&this->nVersion));
+            }
+        }
         nVersion = this->nVersion;
+        if (fOverwintered) {
+            READWRITE(*const_cast<uint32_t*>(&this->nFormatBranchId));
+        }
+        bool isOverwinterV3 = fOverwintered &&
+                              nVersion == 3 &&
+                              nFormatBranchId == 0xDEADBEEF;
+        // FIXME: Replace 0xDEADBEEF with NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId;
         READWRITE(*const_cast<std::vector<CTxIn>*>(&vin));
         READWRITE(*const_cast<std::vector<CTxOut>*>(&vout));
         READWRITE(*const_cast<uint32_t*>(&nLockTime));
+        if (isOverwinterV3) {
+            READWRITE(*const_cast<uint32_t*>(&nExpiryHeight));
+        }
         if (nVersion >= 2) {
```

Comment:
Do we want to restrict this as well (wrt not having soft-forking capacity for Overwinter)? ie:
```
if (isOverwinterV3 || (!fOverwintered && nVersion >= 2)) {
```

Reply:
It might be easier to do this simply by throwing an exception earlier if `fOverwintered && nVersion > 3`.

Reply:
I'm going to remove the exception throwing as a test showed that it would resulted in unexpected code path.  When assigning via `=` GetHash would be invoked which involves serailiazing.

Reply:
Ok I'm not going to remove this anymore based on your later comment.

Reply:
I'm dubious about the exception; are exceptions typically used to signal errors in deserialization?

Reply:
They definitely are in the Bitcoin-inherited (de)serialization code:

- https://github.com/zcash/zcash/blob/master/src/serialize.h#L286-L305
- https://github.com/zcash/zcash/blob/master/src/serialize.h#L455-L457
- https://github.com/zcash/zcash/blob/master/src/serialize.h#L766-L768
  - We added this one.
- https://github.com/zcash/zcash/blob/master/src/main.cpp#L5411-L5417
  - Catches parsing exceptions in network messages.
- https://github.com/zcash/zcash/blob/master/src/core_read.cpp#L100-L105
  - Catches parsing exceptions in hex transactions from RPC method parameters.

I'm of the opinion that an exception is better than having a `C[Mut]Transaction` with undefined state.